### PR TITLE
feat(Pagination): use custom component in pagination

### LINKED ISSUE #1369: Cannot use links in Pagination
UXRFC-439

### PR DESCRIPTION
**Pull request by Arucard89**
Add availability to use custom component in pagination (links, router links, etc.)

(instead of https://github.com/gravity-ui/uikit/pull/2040)

## Summary by Sourcery

Allow Pagination to render its clickable items using a custom root component or anchor links, with per-item props, while preserving accessibility and disabled behavior.

New Features:
- Add `component` prop to Pagination to override the root element for navigation and page buttons, including support for router components and anchors.
- Introduce `getItemProps` callback to supply per-item props such as `href` or `to` for custom Pagination components.

Enhancements:
- Ensure Pagination-managed props override values from `getItemProps` and keep disabled navigation buttons non-interactive buttons.
- Add `buildComponentProps` utility and type helpers to filter user props and support custom button components.
- Expose new Pagination types (`PaginationComponent`, `GetPaginationItemProps`) and Button custom element type for consumers.

Documentation:
- Document the new `component` and `getItemProps` props in both English and Russian Pagination READMEs, including examples for links and router integration.

Tests:
- Expand Pagination unit tests and utils tests to cover custom component rendering, `getItemProps` behavior, accessibility attributes, and disabled navigation edge cases.
- Add Storybook stories demonstrating Pagination usage with custom components and anchor links.

## Summary by Sourcery

Allow Pagination to render its clickable items using a custom root component or anchor links with per-item props while preserving accessibility and disabled behavior.

New Features:
- Add `component` and `getItemProps` props to Pagination to support custom root elements (including router links and anchors) for navigation and page items.
- Export new Pagination-related types (`PaginationComponent`, `GetPaginationItemProps`) and the Button custom element type for consumers.

Enhancements:
- Introduce `buildComponentProps` and `shouldUpdateOnPaginationItemClick` utilities to filter user-supplied props, handle custom components safely, and control when clicks update pagination state.
- Adapt PaginationPage and PaginationButton to use custom components, manage `aria-current` vs `aria-pressed`, and keep disabled navigation items as inert native buttons.
- Add comprehensive unit tests for Pagination and utils covering custom component rendering, link behaviors, and click-handling edge cases.
- Add Storybook stories demonstrating Pagination usage with router-style components and anchor links, including URL-driven initial page handling.

Documentation:
- Extend English and Russian Pagination READMEs to document the new `component` and `getItemProps` APIs with examples for regular links and router integration.

### CODE REVIEW

File: src/components/Pagination/utils.ts [hunk 1]
```diff
@@ -100,3 +110,69 @@ export function getViews({propView, mobile}: {propView: PaginationView; mobile:
...
+    'view',
+    'selected',
+    'disabled',
+    'qa',
+    'aria-current',
+    'extraProps',
+    'children',
+]);
+
+export function buildComponentProps(
```
[the remark this answers is not in the document]
Comment (Arucard89) [reply src/components/Pagination/utils.ts]: I believe it's unnecessary to simplify it further.

File: src/components/Pagination/Pagination.tsx [hunk 2]
```diff
@@ -32,9 +34,27 @@ export const Pagination = ({
     view: propView = 'outlined',
     className,
     qa,
+    component,
+    getItemProps,
 }: PaginationProps) => {
     const mobile = useMobile();
 
+    // Wrap a custom component to strip the `component` prop that `Button` leaks
```
[the remark this answers is not in the document]
Comment (Arucard89) [reply src/components/Pagination/Pagination.tsx]: removed in the last commit

File: src/components/Pagination/__stories__/Pagination.stories.tsx [hunk 3]
```diff
@@ -142,3 +172,101 @@ View.args = {
...
+        if (item.disabled) {
+            return {};
+        }
+        switch (item.action) {
+            case 'first':
+                return {href: getPageHref(1)};
+            case 'previous':
+                return {href: getPageHref(Math.max(1, state.page - 1))};
+            case 'next':
+                return {href: getPageHref(Math.min(lastPage ?? state.page + 1, state.page + 1))};
```
Comment (jhoncool) [inline src/components/Pagination/__stories__/Pagination.stories.tsx]: I don't like that the calculations are happening outside the component. Let's bring back a helper that will do the calculations or something.

File: src/components/Pagination/components/PaginationButton/PaginationButton.tsx [hunk 4]
```diff
@@ -20,6 +21,8 @@ type Props = {
     compact: NonNullable<PaginationProps['compact']>;
     className?: string;
     view: ButtonView;
+    component?: PaginationProps['component'];
```
Comment (jhoncool) [inline src/components/Pagination/components/PaginationButton/PaginationButton.tsx]: It's very common name, maybe `navigationComponent`?

File: src/components/Pagination/components/PaginationButton/PaginationButton.tsx [hunk 5]
```diff
@@ -31,19 +34,28 @@ export const PaginationButton = ({
...
     view,
+    component,
+    getItemProps,
 }: Props) => {
     let button: React.ReactNode = null;
     const {disabled} = item;
     const {t} = i18n.useTranslation();
+    const componentProps = buildComponentProps(component, item, getItemProps);
+    const handleUpdate = (event: React.MouseEvent<HTMLElement>, nextPage: number) => {
+        if (shouldUpdateOnPaginationItemClick(event, Boolean(component))) {
```
Comment (jhoncool) [inline src/components/Pagination/components/PaginationButton/PaginationButton.tsx]: Why do you need to skip `onUpdate`?

File: src/components/Pagination/__stories__/Pagination.stories.tsx [hunk 6]
```diff
@@ -14,6 +14,36 @@ const useState = (args: PaginationProps) => {
...
+
+const FakeRouterLink = React.forwardRef<HTMLAnchorElement, FakeRouterLinkProps>(
+    ({to, children, onClick, ...props}, ref) => (
+        <a
+            ref={ref}
+            {...props}
+            href={to}
+            onClick={(event) => {
+                // Let `Pagination` handle the click first (calls `onUpdate` for plain clicks).
+                onClick?.(event);
```
Comment (jhoncool) [inline src/components/Pagination/__stories__/Pagination.stories.tsx:27]: How it works if you have
```tsx
                <Button
                    {...componentProps}
                    onClick={disabled ? undefined : (event) => handleUpdate(event, 1)}
                >...
```
iside

File: src/components/Pagination/components/PaginationButton/PaginationButton.tsx (hunk 4 above)
Comment (Arucard89) [reply src/components/Pagination/components/PaginationButton/PaginationButton.tsx]: change to `navigationComponent` [e2e012d](https://github.com/gravity-ui/uikit/pull/2701/commits/e2e012d0d4f8e5ccbedf1a53368d2110f31bc282)

File: src/components/Pagination/__stories__/Pagination.stories.tsx (hunk 3 above)
Comment (Arucard89) [reply src/components/Pagination/__stories__/Pagination.stories.tsx]: refactor: [747ae97](https://github.com/gravity-ui/uikit/pull/2701/commits/747ae973fcc39916f95f7430ffe309550cf58926)

File: src/components/Pagination/components/PaginationButton/PaginationButton.tsx (hunk 5 above)
Comment (Arucard89) [reply src/components/Pagination/components/PaginationButton/PaginationButton.tsx]: when a custom component is set, we block onUpdate events for all alternative clicks except the main one (shouldUpdateOnPaginationItemClick). This is necessary so that during alternative clicks (and opening a link in a new tab), the pagination page does not switch on the current tab.

File: src/components/Pagination/__stories__/Pagination.stories.tsx (hunk 6 above)
Comment (Arucard89) [reply src/components/Pagination/__stories__/Pagination.stories.tsx:27]: What happens on click:

**Regular left click on the link:**

1. The browser calls the `<a>` `onClick`, which is handled by `FakeRouterLink`.
2. `FakeRouterLink` first calls `onClick?.(event)`, which is our handler.
3. Since it’s a normal click, our handler calls `onUpdate(4)`, so the current page changes.
4. Then `FakeRouterLink` calls `preventDefault()`, so the browser does not perform a full navigation. This simulates client-side routing.

**Cmd/Ctrl-click to open in a new tab:**

1. Our handler detects the modifier key and does not call `onUpdate`, so the current tab does not change.
2. `FakeRouterLink` does not call `preventDefault()`, so the browser opens `?page=4` in a new tab.

File: src/components/Pagination/__stories__/Pagination.stories.tsx [hunk 7]
```diff
@@ -142,3 +173,79 @@ View.args = {
...
+    const page = getInitialPage(args.page);
+
+    const getItemProps: PaginationProps['getItemProps'] = (_item, itemPage) => ({
+        href: getPageHref(itemPage),
+    });
+
+    return (
+        <Pagination
+            {...args}
+            page={page}
```
Comment (jhoncool) [inline src/components/Pagination/__stories__/Pagination.stories.tsx]: Let's add both examples with prop showInput === true

File: src/components/Pagination/components/PaginationButton/PaginationButton.tsx [hunk 8]
```diff
@@ -31,19 +34,31 @@ export const PaginationButton = ({
...
     compact,
     view,
+    navigationComponent,
+    getItemProps,
 }: Props) => {
     let button: React.ReactNode = null;
     const {disabled} = item;
     const {t} = i18n.useTranslation();
+    const nextPage = getPageNumber(page, item.action);
+    const componentProps = buildComponentProps(navigationComponent, item, getItemProps, nextPage);
```
Comment (jhoncool) [inline src/components/Pagination/components/PaginationButton/PaginationButton.tsx]: Let's use one argument - an object

File: src/components/Pagination/utils.ts [hunk 9]
```diff
@@ -100,3 +110,70 @@ export function getViews({propView, mobile}: {propView: PaginationView; mobile:
...
+    component: PaginationComponent,
+    item: PageItem | ButtonItem,
+    getItemProps?: GetPaginationItemProps,
+    itemPage = 0,
+): Record<string, unknown> {
+    if (!component || (item.type === 'button' && item.disabled)) {
+        return {};
+    }
+
+    const userProps = getItemProps?.(item, itemPage);
```
Comment (jhoncool) [inline src/components/Pagination/utils.ts]: let's use one argument - an object

Comment (Arucard89) [reply src/components/Pagination/utils.ts]: fixed: 09c12e3

File: src/components/Pagination/components/PaginationButton/PaginationButton.tsx (hunk 8 above)
Comment (Arucard89) [reply src/components/Pagination/components/PaginationButton/PaginationButton.tsx]: fixed: 09c12e3

File: src/components/Pagination/__stories__/Pagination.stories.tsx (hunk 7 above)
Comment (Arucard89) [reply src/components/Pagination/__stories__/Pagination.stories.tsx]: add: [9a63c7b](https://github.com/gravity-ui/uikit/pull/2701/commits/9a63c7b5407cff15722b87a740b852c5b4fa3910)

**Comment by jhoncool** [comment]
@korvin89 @amje could you take a look at this PR as well?

**Comment by jhoncool** [comment]
@Arucard89 for me it's okay but let's wait for a review from @korvin89 or @amje

**Review by jhoncool — APPROVED** [review]

**Comment by Arucard89** [comment]
@amje , hi! could you review this PR?

**Comment by amje** [comment]
> @amje , hi! could you review this PR?

@Arucard89 Yes, I'll take a look till the end of the week.

**Comment by Arucard89** [comment]
@amje , hi! Did you manage to view this PR?

File: src/components/Pagination/Pagination.tsx [hunk 10]
```diff
@@ -34,6 +34,8 @@ export const Pagination = (rawProps: PaginationProps) => {
         view: propView = 'outlined',
         className,
         qa,
+        navigationComponent,
+        getItemProps,
```
Comment (amje) [inline src/components/Pagination/Pagination.tsx]: Naming suggestion: `pageComponent` and `getPageProps`. Here "page" refers to a button activating certain page.

File: src/components/Pagination/components/PaginationPage/PaginationPage.tsx [hunk 11]
```diff
@@ -28,16 +39,28 @@ export const PaginationPage = ({item, size, pageSize, className, onUpdate}: Prop
...
     return (
         <Button
+            {...componentProps}
             size={size}
             key={view}
             view={view}
             selected={item.current}
             className={className}
-            onClick={() => onUpdate(item.page, pageSize)}
+            onClick={(event) => {
```
Comment (amje) [inline src/components/Pagination/components/PaginationPage/PaginationPage.tsx]: Why don't you get onClick and aria-* attributes from the `buildComponentProps` utility? There is a bit of duplication across some components

File: src/components/Pagination/index.ts [hunk 12]
```diff
@@ -1,2 +1,8 @@
 export * from './Pagination';
-export type {PaginationProps} from './types';
+export type {
+    ButtonItem as PaginationButtonItem,
+    GetPaginationItemProps,
```
Comment (amje) [inline src/components/Pagination/index.ts]: Better be named with component prefix `PaginationSmthPropsGetter`

File: src/components/Pagination/components/PaginationPage/PaginationPage.tsx (hunk 11 above)
Comment (Arucard89) [reply src/components/Pagination/components/PaginationPage/PaginationPage.tsx]: Moved onClick and aria-current into buildComponentProps — it now takes onUpdate/pageSize, and both components just spread the result. onClick is computed outside the component check so default pagination (without a custom component) keeps firing onUpdate.


https://github.com/gravity-ui/uikit/pull/2701/commits/1d61b8ab5788892be3a29f02bf3cc07ef39bac4a

File: src/components/Pagination/Pagination.tsx (hunk 10 above)
Comment (Arucard89) [reply src/components/Pagination/Pagination.tsx]: renamed: https://github.com/gravity-ui/uikit/pull/2701/commits/1d61b8ab5788892be3a29f02bf3cc07ef39bac4a

File: src/components/Pagination/index.ts (hunk 12 above)
Comment (Arucard89) [reply src/components/Pagination/index.ts]: renamed: https://github.com/gravity-ui/uikit/pull/2701/commits/1d61b8ab5788892be3a29f02bf3cc07ef39bac4a

File: src/components/Pagination/index.ts [hunk 13]
```diff
@@ -1,2 +1,8 @@
 export * from './Pagination';
-export type {PaginationProps} from './types';
+export type {
+    ButtonItem as PaginationButtonItem,
+    PaginationPagePropsGetter,
+    PageItem as PaginationPageItem,
+    PaginationComponent,
```
Comment (amje) [inline src/components/Pagination/index.ts]: `PaginationPageComponent` is more accurate for that. `PaginationComponent` is misleading

**Review by amje — CHANGES_REQUESTED** [review]

File: src/components/Pagination/index.ts (hunk 13 above)
Comment (Arucard89) [reply src/components/Pagination/index.ts]: renamed: https://github.com/gravity-ui/uikit/pull/2701/commits/6fc742eccd90e6ca23bf6af1b9b27bd80d312c0c

**Review by amje — APPROVED** [review]

**Review by jhoncool — APPROVED** [review]
